### ISSUE TITLE
Error when trying to connect to an Appium server in a Selenium Grid

### ISSUE THREAD

**Issue by GhenTerr** [appium/appium#8492]
## The problem

I got Error code while I was trying to run tests through Selenium grid, the driver was created and app opened but shortly after I got a status code 500 error. I'm using Ruby client.

## Environment

* Appium version (or git revision) that exhibits the issue: 1.6.4
* Desktop OS/version used to run Appium: Windows 10 1607
* Node.js version (unless using Appium.app|exe): 6.10.0
* Mobile platform/version under test: Android 6.0.1
* Real device or emulator/simulator: Samsung Galaxy A5 2017
* Appium CLI or Appium.app|exe: CLI

## Details

I tried with Node JS client, and there seems to be no problem. However the only thing I did is creating then quiting a driver. 

## Link to Appium logs

https://gist.github.com/GhenTerr 

## Code To Reproduce Issue [ Good To Have ]

All the code does it open the message app and click the new message button, that's it

```
require 'appium_lib'

desired_caps = {
  caps:       {
    platformName: 'Android',
    platformVersion: '6.0.1',
    deviceName: 'Galaxy A5 (2017)',
    appPackage: 'com.samsung.android.messaging',
    appActivity: 'com.android.mms.ui.ConversationComposer'
  },
  appium_lib: {
    server_url: 'http://192.168.43.78:4444/wd/hub'
  }
}

driver = Appium::Driver.new(desired_caps)
driver.start_driver()
driver.find_element(:xpath, '//*[@resource-id="com.samsung.android.messaging:id/floating_action_button"]').click();

driver.driver_quit
```
Also the config json

```
`{
    "capabilities": [{
        "browserName": "Galaxy A5 (2017)",
        "version": "6.0.1",
        "platform": "Android",
        "maxInstances": 1
    }],
    "configuration": {
        "cleanUpCycle": 2000,
        "timeout": 30000,
        "proxy": "org.openqa.grid.selenium.proxy.DefaultRemoteProxy",
        "url": "http://127.0.0.1:5000/wd/hub",
        "host": "127.0.0.1",
        "port": 5000,
        "maxSession": 1,
        "register": true,
        "registerCycle": 5000,
        "hubPort": 4444,
        "hubHost": "127.0.0.1"
    }
}`
```

**Comment by mykola-mokhnach** [comment]
@KazuCocoa do you know any possible reason?

**Comment by KazuCocoa** [comment]
According to https://gist.github.com/GhenTerr/6c93a207cb788d18a1940c5e818a4c0c, the error was caused selenium-webdriver layer.
Could you try ruby_lib 8.0.2?
The version uses selenium-webdriver2.x. If 8.0.2 works well, the error is selenium-webdriver 3.x and ruby_lib related issue.

I'll investigate the issue, though.

**Comment by KazuCocoa** [comment]
What about appending `browserName` for Ruby script?

```diff
desired_caps = {
  caps:       {
+    browserName: "Galaxy A5 (2017)",
    platformName: 'Android',
    platformVersion: '6.0.1',
    deviceName: 'Galaxy A5 (2017)',
    appPackage: 'com.samsung.android.messaging',
    appActivity: 'com.android.mms.ui.ConversationComposer'
  },
  appium_lib: {
    server_url: 'http://192.168.43.78:4444/wd/hub'
  }
}
```

**Comment by KazuCocoa** [comment]
I think https://github.com/appium/ruby_lib/issues/541#issuecomment-302916016 is a simmiler issue and I've been investigating yet.
If this issue will fix with https://github.com/appium/appium/issues/8492#issuecomment-302878672, I'm able to fix in `ruby_lib` side. If not, I have to investigate more...

**Comment by KazuCocoa** [comment]
I summarize my above comments.

1. Could you try https://github.com/KazuCocoa/ruby_lib/tree/require_over_selenium_webdriver2 with selenium-webdriver2.x? https://github.com/appium/appium/issues/8492#issuecomment-302916204
2. Could you try ruby_lib 8.0.2?
3. What about appending browserName and version for Ruby script? https://github.com/appium/appium/issues/8492#issuecomment-302878672

I wish grid works with the 1st solution...

**Comment by KazuCocoa** [comment]
progress: Failed to conduct tests from `appium_lib9.2.0`, I guess.

**Comment by GhenTerr** [comment]
Switching to ruby Appium_lib 8.0.2 seems to fixed the problem!

**Comment by KazuCocoa** [comment]
Pretty good.

Could you set  `browserName` and `version` for desired_caps's `:cap` such as https://github.com/appium/ruby_lib/blob/master/grid/appium.txt.ios.example#L5-L6 ?
I can work with https://github.com/appium/ruby_lib/tree/master/grid and ruby_lib9.4.4.

**Comment by GhenTerr** [comment]
I tried with the following `caps` along ruby lib 9.4.4 and got the same error

```
  caps:       {
    browserName: 'Galaxy A5 (2017)',
    version: '6.0.1',
    platformName: 'Android',
    platformVersion: '6.0.1',
    deviceName: 'Galaxy A5 (2017)',
    appPackage: 'com.samsung.android.messaging',
    appActivity: 'com.android.mms.ui.ConversationComposer'
  }
```

**Comment by KazuCocoa** [comment]
@GhenTerr 

Thanks for checking, I appreciate your help to investigate the issue.
I found the following issue.

- Grid does not handle w3c capabilities correctly
     - https://github.com/SeleniumHQ/selenium/issues/3808

Could you try [2.x](http://selenium-release.storage.googleapis.com/index.html?path=2.53/) as grid server and the following capability?

```
  caps:       {
    browserName: 'Galaxy A5 (2017)',
    platformName: 'Android',
    platformVersion: '6.0.1',
    deviceName: 'Galaxy A5 (2017)',
    appPackage: 'com.samsung.android.messaging',
    appActivity: 'com.android.mms.ui.ConversationComposer'
  }
```

If creating a session is failed in this case, I have to check handshake between the grid and the client, I think.

**Comment by GhenTerr** [comment]
No problem!

I got the same error again with Selenium grid server 2.53.1 with same Ruby lib and Appium version.

**Comment by KazuCocoa** [comment]
Thanks!
hm...

The following branch didn't work (means the same issue was caused), did you?

> 1. Could you try https://github.com/KazuCocoa/ruby_lib/tree/require_over_selenium_webdriver2 with selenium-webdriver2.x(the client side)?

from https://github.com/appium/appium/issues/8492#issuecomment-302937817

If the above branch works well, I think this issue is selenium-webdriver related issue or capability issue appium_lib generate.

**Comment by GhenTerr** [comment]
Just tried that branch, it indeed didn't work.

**Comment by KazuCocoa** [comment]
Thanks for your update.

The error is same as https://github.com/appium/ruby_lib/issues/541 . I need more time to investigate the issue.

**Comment by KazuCocoa** [comment]
https://github.com/appium/ruby_lib/pull/600 will fix 500 status code error.

**Comment by KazuCocoa** [comment]
just published 9.4.7, fixed appium/ruby_lib#600
Could you try 9.4.7 if you have a time? @GhenTerr

**Comment by GhenTerr** [comment]
It seems to be fixed! I tested it with Selenium server 3.2.0 and 3.4.0. Thanks!
